<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.20.1 by Michael Rose
  Copyright 2013-2019 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Performing clustering using FEATS - Edwin Vans</title>
<meta name="description" content="This notebook shows how to perform clustering using the FEATS package.">


  <meta name="author" content="Edwin Vans">


<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Edwin Vans">
<meta property="og:title" content="Performing clustering using FEATS">
<meta property="og:url" content="http://localhost:4000/performing-clustering-using-feats/">


  <meta property="og:description" content="This notebook shows how to perform clustering using the FEATS package.">







  <meta property="article:published_time" content="2020-09-05T00:00:00+12:00">





  

  


<link rel="canonical" href="http://localhost:4000/performing-clustering-using-feats/">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": "Edwin Vans",
      "url": "http://localhost:4000/"
    
  }
</script>






<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Edwin Vans Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          EDWIN VANS
          <span class="site-subtitle">.com</span>
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/about/">About</a>
            </li><li class="masthead__menu-item">
              <a href="/projects/">Projects</a>
            </li></ul>
        
        <button class="search__toggle" type="button">
          <span class="visually-hidden">Toggle search</span>
          <svg class="icon" width="16" height="16" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 15.99 16">
            <path d="M15.5,13.12L13.19,10.8a1.69,1.69,0,0,0-1.28-.55l-0.06-.06A6.5,6.5,0,0,0,5.77,0,6.5,6.5,0,0,0,2.46,11.59a6.47,6.47,0,0,0,7.74.26l0.05,0.05a1.65,1.65,0,0,0,.5,1.24l2.38,2.38A1.68,1.68,0,0,0,15.5,13.12ZM6.4,2A4.41,4.41,0,1,1,2,6.4,4.43,4.43,0,0,1,6.4,2Z" transform="translate(-.01)"></path>
          </svg>
        </button>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  
  <div class="sidebar sticky">
  


<div itemscope itemtype="https://schema.org/Person">

  

  <div class="author__content">
    
      <h3 class="author__name" itemprop="name">Edwin Vans</h3>
    
    
      <div class="author__bio" itemprop="description">
        <p>Hello, Iâ€™m Edwin Vans. An academic by profession, who loves to write code.</p>

      </div>
    
  </div>

  <div class="author__urls-wrapper">
    <button class="btn btn--inverse">Follow</button>
    <ul class="author__urls social-icons">
      
        <li itemprop="homeLocation" itemscope itemtype="https://schema.org/Place">
          <i class="fas fa-fw fa-map-marker-alt" aria-hidden="true"></i> <span itemprop="name">Suva, Fiji</span>
        </li>
      

      
        
          
            <li><a href="mailto:info@edwinvans.com" rel="nofollow noopener noreferrer"><i class="fas fa-fw fa-envelope-square" aria-hidden="true"></i><span class="label">Email</span></a></li>
          
        
          
        
          
        
          
        
          
            <li><a href="https://github.com/edwinv87" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i><span class="label">GitHub</span></a></li>
          
        
          
        
          
            <li><a href="https://www.linkedin.com/in/edwin-vans-4067aa62/" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span class="label">LinkedIn</span></a></li>
          
        
      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <!--
  <li>
    <a href="http://link-to-whatever-social-network.com/user/" itemprop="sameAs" rel="nofollow noopener noreferrer">
      <i class="fas fa-fw" aria-hidden="true"></i> Custom Social Profile Link
    </a>
  </li>
-->
    </ul>
  </div>
</div>

  
  </div>



  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="Performing clustering using FEATS">
    <meta itemprop="description" content="This notebook shows how to perform clustering using the FEATS package.">
    <meta itemprop="datePublished" content="2020-09-05T00:00:00+12:00">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">Performing clustering using FEATS
</h1>
          


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        5 minute read
      
    

  </p>

        </header>
      

      <section class="page__content" itemprop="text">
        
        <p>This notebook shows how to perform clustering using the FEATS package.</p>

<h2 id="import-libraries">Import libraries</h2>

<p>The first step is to import libraries including functions from the FEATS package. We use the <code class="language-plaintext highlighter-rouge">json</code> package to load a python dictionary of dataset specific parameters stored as a text file. The <code class="language-plaintext highlighter-rouge">pandas</code> package is used to open data files stored as csv and to create dataframes. The <code class="language-plaintext highlighter-rouge">adjusted_rand_score</code> function from the scikit-learn package is used for computing ARI. The <code class="language-plaintext highlighter-rouge">SingleCell</code> class from the <a href="https://github.com/edwinv87/singlecell">singlecelldata</a> package is used for easy management of single-cell datasets. The functions in FEATS pcakage accepts SingleCell objects only to perform analysis. Plotting functions from the <a href="https://github.com/edwinv87/scplotlib">scplotlib</a> library are used to generate cluster visualizations. The Cluster function from FEATS package is used to perform clustering.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Import core libraries
</span><span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics.cluster</span> <span class="kn">import</span> <span class="n">adjusted_rand_score</span>

<span class="c1"># Import SingleCell
</span><span class="kn">from</span> <span class="nn">singlecelldata</span> <span class="kn">import</span> <span class="n">SingleCell</span>

<span class="c1"># Import visualization functions
</span><span class="kn">from</span> <span class="nn">scplotlib</span> <span class="kn">import</span> <span class="n">PCAPlot</span><span class="p">,</span> <span class="n">SilhouettePlot</span><span class="p">,</span> <span class="n">GeneExpHeatmap</span>

<span class="c1"># Import the FEATS methods
</span><span class="kn">from</span> <span class="nn">feats</span> <span class="kn">import</span> <span class="n">Cluster</span><span class="p">,</span> <span class="n">GeneFilter</span><span class="p">,</span> <span class="n">LogFilter</span>

</code></pre></div></div>

<h2 id="load-dataset">Load dataset</h2>

<p>The next step is to load the dataset. Here we use Pandas to import data stored in csv format into dataframes. The dataframes are then used to create the SingleCell object as the functions in the FEATS package accepts SingleCell objects.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Python dictionary storing dataset specific parameters.
</span><span class="n">dataset_params</span> <span class="o">=</span> <span class="n">json</span><span class="p">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">"dset_param.txt"</span><span class="p">))</span>

<span class="n">dataset</span> <span class="o">=</span> <span class="s">"kolodziejczyk"</span>
<span class="n">dset_name</span> <span class="o">=</span> <span class="n">dataset_params</span><span class="p">[</span><span class="n">dataset</span><span class="p">][</span><span class="s">'dset_name'</span><span class="p">]</span>
<span class="n">path</span> <span class="o">=</span> <span class="s">"datasets/"</span>
<span class="n">data_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="n">dset_name</span> <span class="o">+</span> <span class="s">'/'</span> <span class="o">+</span> <span class="n">dset_name</span> <span class="o">+</span> <span class="s">"_data.csv"</span>
<span class="n">celldata_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="n">dset_name</span> <span class="o">+</span> <span class="s">'/'</span> <span class="o">+</span> <span class="n">dset_name</span> <span class="o">+</span> <span class="s">"_celldata.csv"</span>
<span class="n">genedata_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="n">dset_name</span> <span class="o">+</span> <span class="s">'/'</span> <span class="o">+</span> <span class="n">dset_name</span> <span class="o">+</span> <span class="s">"_genedata.csv"</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">celldata</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">celldata_path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">genedata</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genedata_path</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

<span class="c1"># Create a single cell object
</span><span class="n">sc</span> <span class="o">=</span> <span class="n">SingleCell</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span><span class="n">data</span><span class="p">,</span><span class="n">celldata</span><span class="p">,</span><span class="n">genedata</span><span class="p">)</span>

<span class="c1"># Get parameters from the dictionary
</span><span class="n">label</span> <span class="o">=</span> <span class="n">dataset_params</span><span class="p">[</span><span class="n">dataset</span><span class="p">][</span><span class="s">'label'</span><span class="p">]</span> <span class="c1"># This is the name of the
</span>                                         <span class="c1"># column which stores true cell labels.
</span><span class="n">nc</span> <span class="o">=</span> <span class="n">dataset_params</span><span class="p">[</span><span class="n">dataset</span><span class="p">][</span><span class="s">'nc'</span><span class="p">]</span> <span class="c1"># This is the true number of
</span>                                   <span class="c1"># clusters in the dataset.
</span>
<span class="c1"># sc.print()
</span></code></pre></div></div>

<h2 id="log-transformation">Log transformation</h2>

<p>The <code class="language-plaintext highlighter-rouge">LogFilter</code> function takes the Log (with base 2) of the expression counts stored in sc and stores the log transformed values in the data assay.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Take a log transformation of gene expressions
</span><span class="n">sc</span> <span class="o">=</span> <span class="n">LogFilter</span><span class="p">(</span><span class="n">sc</span><span class="p">)</span>
</code></pre></div></div>

<h2 id="gene-filtering">Gene filtering</h2>

<p>The <code class="language-plaintext highlighter-rouge">GeneFilter</code> function filters out genes based on the filtering criteria. It accepts three arguments apart from the sc object which are <code class="language-plaintext highlighter-rouge">min_cells</code>, <code class="language-plaintext highlighter-rouge">max_cells</code> and <code class="language-plaintext highlighter-rouge">expr_thres</code>. The GeneFilter function filters out genes expressed (with expression value less than or equal to <code class="language-plaintext highlighter-rouge">expr_thres</code>) in less than <code class="language-plaintext highlighter-rouge">min_cells</code> and greater than <code class="language-plaintext highlighter-rouge">max_cells</code> number of cells. Here we use 10% of cells as the <code class="language-plaintext highlighter-rouge">min_cells</code> and 90% of cells as <code class="language-plaintext highlighter-rouge">max_cells</code>. <code class="language-plaintext highlighter-rouge">expr_thres</code> is 0 by default.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Perform Gene Filtering
</span><span class="n">sc</span> <span class="o">=</span> <span class="n">GeneFilter</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span>
                <span class="n">min_cells</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="mi">10</span><span class="o">/</span><span class="mi">100</span><span class="p">)</span> <span class="o">*</span> <span class="n">sc</span><span class="p">.</span><span class="n">dim</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="n">max_cells</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="mi">90</span><span class="o">/</span><span class="mi">100</span><span class="p">)</span> <span class="o">*</span> <span class="n">sc</span><span class="p">.</span><span class="n">dim</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                <span class="n">expr_thres</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Applying Gene Filter . . .
Number of features remaining after gene filtering:  10725
</code></pre></div></div>

<h2 id="clustering">Clustering</h2>

<p>The <code class="language-plaintext highlighter-rouge">Cluster</code> function from FEATS is used to cluster the cells in the dataset. The first argument is the SingleCell object. The second argument is the number of clusters in the data <code class="language-plaintext highlighter-rouge">n_clusters</code>. Since we know this for this dataset, we pass the true number of clusters, <code class="language-plaintext highlighter-rouge">nc</code>. This parameter also accepts a python list of integers as number of clusters. Clustering will be performed using all the integers in the list. It also accepts the string <code class="language-plaintext highlighter-rouge">'gap'</code>, in which case it will estimate the number of clusters using gap statistic. The output argument is the SingleCell object which stores the computed clusters and other information. The second output argument is the number of clusters in the data (which we donâ€™t need in this case).</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Perform clustering
</span><span class="n">sc</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">Cluster</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="n">nc</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Computing Temporary Clusters . . .
Performing Feature Selection . . .
Computing 3 clusters using best features . . .
Saving final cluster labels in Single Cell object . . .
</code></pre></div></div>

<h2 id="adjusted-rand-index">Adjusted rand index</h2>

<p>We can assess the clustering performance by computing the ARI using <code class="language-plaintext highlighter-rouge">adjusted_rand_score</code> function from scikit-learn package. The sc object stores the true cluster labels. We can get a numeric version of the true cell labels using the function <code class="language-plaintext highlighter-rouge">getNumericCellLabels</code> if it is stored as a string or in any other form. The FEATS clustering also stores the computed cluster labels in the celldata assay under the column name <code class="language-plaintext highlighter-rouge">FEATS_k_Clusters</code>, where <code class="language-plaintext highlighter-rouge">k</code> is the number of clusters.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Compute and print the Adjusted Rand Score
</span><span class="k">print</span> <span class="p">(</span><span class="s">"Adjusted Rand Index: "</span><span class="p">)</span>
<span class="k">print</span> <span class="p">(</span><span class="n">adjusted_rand_score</span><span class="p">(</span><span class="n">sc</span><span class="p">.</span><span class="n">getNumericCellLabels</span><span class="p">(</span><span class="n">label</span><span class="p">),</span> <span class="n">sc</span><span class="p">.</span><span class="n">getCellData</span><span class="p">(</span><span class="s">"FEATS_3_Clusters"</span><span class="p">)))</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Adjusted Rand Index: 
0.9919841315168355
</code></pre></div></div>

<h2 id="visualize-the-clusters-formed">Visualize the clusters formed</h2>

<p>We can visualize the clusters formed using FEATS in various ways. (1) We can do a PCA plot which is a scatter plot in 2D after reducing the dimensionality using PCA. The scatter plot can be colored using the <code class="language-plaintext highlighter-rouge">color_by</code> argument. We color the plot according to clusters computed by FEATS (fig1). (2) We can do a silhouette plot which will show the silhouette scores of the samples in each cluster (sorted by cluster and silhouette score). We have to specify the cluster information using the <code class="language-plaintext highlighter-rouge">cluster_by</code> argument. Silhouette score is a measure of how similar cells are to their own cluster compared to other clusters. The silhouette values are in the range $[-1, 1]$. Values closer to 1 indicate that samples are well matched to their own cluster and poorly matched to neighbouring clusters (fig2).</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">fig1</span> <span class="o">=</span> <span class="n">PCAPlot</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span>
               <span class="n">color_by</span> <span class="o">=</span> <span class="s">"FEATS_3_Clusters"</span><span class="p">,</span>
               <span class="n">marker_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
               <span class="n">marker_thickness</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">fig1</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<p><img src="/assets/images/fig1.png" alt="Fig 1" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">fig2</span> <span class="o">=</span> <span class="n">SilhouettePlot</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">cluster_by</span> <span class="o">=</span> <span class="s">"FEATS_3_Clusters"</span><span class="p">)</span>
<span class="n">fig2</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<p><img src="/assets/images/fig2.png" alt="Fig 2" /></p>

<h2 id="high-scoring-genes-top-genes">High scoring genes (Top genes)</h2>

<p>We can visualize the top genes according to their F-scores. To do this, we can generate a heatmap of the genes sorted by the F-scores using the <code class="language-plaintext highlighter-rouge">GeneExpHeatmap</code> function from scplotlib package (fig3). Cells can be sorted according to their clusters with a horizontal colored bar on top of the heatmap indicating the cluster. The <code class="language-plaintext highlighter-rouge">color_by</code> argument is used to do this. The genes can be sorted also using the <code class="language-plaintext highlighter-rouge">sort_by</code> argument. We use the F-scores stored in the sc object after performing FEATS clustering. The F-scores are stored under the column <code class="language-plaintext highlighter-rouge">FEATS_F_Score</code> in the genedata assay. We can also point to the column which stores the gene names using the <code class="language-plaintext highlighter-rouge">name_by</code> argument. In this case the column <code class="language-plaintext highlighter-rouge">feature_symbol</code> stores the gene names in the genedata assay. In addition, we can specify the number of genes we want using the <code class="language-plaintext highlighter-rouge">top_num_genes</code> argument. We can also generate the heatmap where cells are sorted according to the temporary clusters computed by FEATS (fig4). The temporary clusters are computed to perform ANOVA.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Plot top expressions
</span><span class="n">fig3</span> <span class="o">=</span> <span class="n">GeneExpHeatmap</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span>
                      <span class="n">color_by</span> <span class="o">=</span> <span class="s">'FEATS_3_Clusters'</span><span class="p">,</span>
                      <span class="n">sort_by</span> <span class="o">=</span> <span class="s">'FEATS_F_Score'</span><span class="p">,</span>
                      <span class="n">name_by</span> <span class="o">=</span>  <span class="s">'feature_symbol'</span><span class="p">,</span>
                      <span class="n">top_num_genes</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">fig3</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<p><img src="/assets/images/fig3.png" alt="Fig 3" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">fig4</span> <span class="o">=</span> <span class="n">GeneExpHeatmap</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span>
                      <span class="n">color_by</span> <span class="o">=</span> <span class="s">'FEATS_3_Clusters_Temp'</span><span class="p">,</span>
                      <span class="n">sort_by</span> <span class="o">=</span> <span class="s">'FEATS_F_Score'</span><span class="p">,</span>
                      <span class="n">name_by</span> <span class="o">=</span>  <span class="s">'feature_symbol'</span><span class="p">,</span>
                      <span class="n">top_num_genes</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">fig4</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<p><img src="/assets/images/fig4.png" alt="Fig 4" /></p>

<h3 id="contact">Contact</h3>

<p>For clarification on any of the functions used in this document please contact us via email: vans.edw@gmail.com</p>

        
      </section>

      <footer class="page__meta">
        
        


        
  <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2020-09-05T00:00:00+12:00">September 5, 2020</time></p>


      </footer>

      <section class="page__share">
  

  <a href="https://twitter.com/intent/tweet?text=Performing+clustering+using+FEATS%20http%3A%2F%2Flocalhost%3A4000%2Fperforming-clustering-using-feats%2F" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=http%3A%2F%2Flocalhost%3A4000%2Fperforming-clustering-using-feats%2F" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A4000%2Fperforming-clustering-using-feats%2F" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="/performing-batch-correction-using-feats/" class="pagination--pager" title="Performing batch correction using FEATS
">Previous</a>
    
    
      <a href="/performing-outlier-detection-using-feats/" class="pagination--pager" title="Performing outlier detection using FEATS
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
    <div class="page__related">
      <h4 class="page__related-title">You May Also Enjoy</h4>
      <div class="grid__wrapper">
        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/performing-outlier-detection-using-feats/" rel="permalink">Performing outlier detection using FEATS
</a>
      
    </h2>
    


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        3 minute read
      
    

  </p>

    <p class="archive__item-excerpt" itemprop="description">In this notebook, we perform outlier analysis using FEATS on mouse intestine dataset.

</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/performing-batch-correction-using-feats/" rel="permalink">Performing batch correction using FEATS
</a>
      
    </h2>
    


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        4 minute read
      
    

  </p>

    <p class="archive__item-excerpt" itemprop="description">In this code example, we describe how to perform batch correction and integration using FEATS. We will use an artificial dataset, however, the same steps can...</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/welcome-to-my-site/" rel="permalink">Welcome to my site!
</a>
      
    </h2>
    


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        less than 1 minute read
      
    

  </p>

    <p class="archive__item-excerpt" itemprop="description">Hello and welcome to my site! My name is Edwin Vans. Iâ€™m an academic who loves to write code and work on software and hardware projects. Stay tuned to my blo...</p>
  </article>
</div>

        
      </div>
    </div>
  
  
</div>

    </div>

    
      <div class="search-content">
        <div class="search-content__inner-wrap"><form class="search-content__form" onkeydown="return event.key != 'Enter';">
    <label class="sr-only" for="search">
      Enter your search term...
    </label>
    <input type="search" id="search" class="search-input" tabindex="-1" placeholder="Enter your search term..." />
  </form>
  <div id="results" class="results"></div></div>

      </div>
    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    

    
      
        
      
        
      
        
          <li><a href="https://github.com/edwinv87" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
        
      
        
      
        
      
        
      
    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2020 Edwin Vans. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>




<script src="/assets/js/lunr/lunr.min.js"></script>
<script src="/assets/js/lunr/lunr-store.js"></script>
<script src="/assets/js/lunr/lunr-en.js"></script>







  </body>
</html>
