<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.20.1 by Michael Rose
  Copyright 2013-2019 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Performing batch correction using FEATS - Edwin Vans</title>
<meta name="description" content="In this code example, we describe how to perform batch correction and integration using FEATS. We will use an artificial dataset, however, the same steps can be followed for any dataset.">


  <meta name="author" content="Edwin Vans">


<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Edwin Vans">
<meta property="og:title" content="Performing batch correction using FEATS">
<meta property="og:url" content="http://localhost:4000/performing-batch-correction-using-feats/">


  <meta property="og:description" content="In this code example, we describe how to perform batch correction and integration using FEATS. We will use an artificial dataset, however, the same steps can be followed for any dataset.">







  <meta property="article:published_time" content="2020-09-05T00:00:00+12:00">





  

  


<link rel="canonical" href="http://localhost:4000/performing-batch-correction-using-feats/">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": "Edwin Vans",
      "url": "http://localhost:4000/"
    
  }
</script>






<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Edwin Vans Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          EDWIN VANS
          <span class="site-subtitle">.com</span>
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/about/">About</a>
            </li><li class="masthead__menu-item">
              <a href="/projects/">Projects</a>
            </li><li class="masthead__menu-item">
              <a href="/consultancy/">Consultancy</a>
            </li></ul>
        
        <button class="search__toggle" type="button">
          <span class="visually-hidden">Toggle search</span>
          <svg class="icon" width="16" height="16" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 15.99 16">
            <path d="M15.5,13.12L13.19,10.8a1.69,1.69,0,0,0-1.28-.55l-0.06-.06A6.5,6.5,0,0,0,5.77,0,6.5,6.5,0,0,0,2.46,11.59a6.47,6.47,0,0,0,7.74.26l0.05,0.05a1.65,1.65,0,0,0,.5,1.24l2.38,2.38A1.68,1.68,0,0,0,15.5,13.12ZM6.4,2A4.41,4.41,0,1,1,2,6.4,4.43,4.43,0,0,1,6.4,2Z" transform="translate(-.01)"></path>
          </svg>
        </button>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  
  <div class="sidebar sticky">
  


<div itemscope itemtype="https://schema.org/Person">

  
    <div class="author__avatar">
      
        <img src="/assets/images/edwin.jpg" alt="Edwin Vans" itemprop="image">
      
    </div>
  

  <div class="author__content">
    
      <h3 class="author__name" itemprop="name">Edwin Vans</h3>
    
    
  </div>

  <div class="author__urls-wrapper">
    <button class="btn btn--inverse">Follow</button>
    <ul class="author__urls social-icons">
      
        <li itemprop="homeLocation" itemscope itemtype="https://schema.org/Place">
          <i class="fas fa-fw fa-map-marker-alt" aria-hidden="true"></i> <span itemprop="name">Suva, Fiji</span>
        </li>
      

      
        
          
            <li><a href="mailto:info@edwinvans.com" rel="nofollow noopener noreferrer"><i class="fas fa-fw fa-envelope-square" aria-hidden="true"></i><span class="label">Email</span></a></li>
          
        
          
        
          
        
          
        
          
            <li><a href="https://github.com/edwinv87" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i><span class="label">GitHub</span></a></li>
          
        
          
        
          
            <li><a href="https://www.linkedin.com/in/edwin-vans-4067aa62/" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span class="label">LinkedIn</span></a></li>
          
        
      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <!--
  <li>
    <a href="http://link-to-whatever-social-network.com/user/" itemprop="sameAs" rel="nofollow noopener noreferrer">
      <i class="fas fa-fw" aria-hidden="true"></i> Custom Social Profile Link
    </a>
  </li>
-->
    </ul>
  </div>
</div>

  
  </div>



  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="Performing batch correction using FEATS">
    <meta itemprop="description" content="In this code example, we describe how to perform batch correction and integration using FEATS. We will use an artificial dataset, however, the same steps can be followed for any dataset.">
    <meta itemprop="datePublished" content="2020-09-05T00:00:00+12:00">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">Performing batch correction using FEATS
</h1>
          


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        4 minute read
      
    

  </p>

        </header>
      

      <section class="page__content" itemprop="text">
        
        <p>In this code example, we describe how to perform batch correction and integration using FEATS. We will use an artificial dataset, however, the same steps can be followed for any dataset.</p>

<h2 id="import-libraries">Import libraries</h2>

<p>Firstly, we need to import libraries to enable us to use the modules. For batch correction, we import <code class="language-plaintext highlighter-rouge">CorrectBatches</code> which is going to perform correction or allignment, and <code class="language-plaintext highlighter-rouge">IntegrateBatches</code> which is going to find common genes among the datasets and combine/merge all the datasets into one. For more information on these functions please use <code class="language-plaintext highlighter-rouge">help()</code> command by passing the function name as the argument, e.g., <code class="language-plaintext highlighter-rouge">help(IntegrateBatches)</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Import core libraries
</span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>

<span class="c1"># Import our libraries
</span><span class="kn">from</span> <span class="nn">singlecelldata</span> <span class="kn">import</span> <span class="n">SingleCell</span>
<span class="kn">from</span> <span class="nn">feats</span> <span class="kn">import</span> <span class="n">CorrectBatches</span><span class="p">,</span> <span class="n">IntegrateBatches</span>

<span class="c1"># Visualization
</span><span class="kn">from</span> <span class="nn">scplotlib</span> <span class="kn">import</span> <span class="n">tSNEPlot</span>
</code></pre></div></div>

<h2 id="import-datasets">Import datasets</h2>

<p>The second step is to import the datasets into the notebook environment. To do this we will use the Pandas <code class="language-plaintext highlighter-rouge">read_csv</code> function to read the data from csv format files into dataframes. Here we have two batches, batch1 and batch2. Apart from the gene expression matrix we will also import cell data which consists of the cell type information. Note that cell type information is arbitrarily used here.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Import Data
</span>
<span class="n">batch1</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"data/batch1.csv"</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">batch2</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"data/batch2ii.csv"</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">b1_celldata</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"data/batch1_labels.csv"</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">b1_celldata</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">b1_celldata</span><span class="p">.</span><span class="n">values</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s">"cell type"</span><span class="p">])</span>

<span class="n">b2_celldata</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"data/batch2ii_labels.csv"</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">b2_celldata</span> <span class="o">=</span> <span class="n">b2_celldata</span><span class="p">.</span><span class="n">T</span>
<span class="n">b2_celldata</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">b2_celldata</span><span class="p">.</span><span class="n">values</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s">"cell type"</span><span class="p">])</span>
</code></pre></div></div>

<h2 id="create-singlecell-objects-from-the-data">Create SingleCell objects from the data</h2>

<p>Next, we will create SingleCell objects from the data using the <code class="language-plaintext highlighter-rouge">SingleCell</code> class. Note here that we are only passing the data and  the celldata dataframes into the class constructor. We are not passing any genedata dataframe. The constructor will automatically create a genedata dataframe and store gene names as ‘gene1’, ‘gene2’, …, ‘gened’ under the data column ‘gene_names’. We can call the <code class="language-plaintext highlighter-rouge">print()</code> method to see a summary of the SingleCell object.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Create a single cell object  
</span>
<span class="n">bat1</span> <span class="o">=</span> <span class="n">SingleCell</span><span class="p">(</span><span class="n">dataset</span> <span class="o">=</span> <span class="s">"simulation1"</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">batch1</span><span class="p">,</span> <span class="n">celldata</span> <span class="o">=</span> <span class="n">b1_celldata</span><span class="p">)</span>
<span class="n">bat2</span> <span class="o">=</span> <span class="n">SingleCell</span><span class="p">(</span><span class="n">dataset</span> <span class="o">=</span> <span class="s">"simulation2"</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">batch2</span><span class="p">,</span> <span class="n">celldata</span> <span class="o">=</span> <span class="n">b2_celldata</span><span class="p">)</span>

<span class="n">bat2</span><span class="p">.</span><span class="k">print</span><span class="p">()</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>------------------------------------------------------------------------------
Dataset:  simulation2
------------------------------------------------------------------------------
Dimension:  (100, 1000)
Cell Metadata:  ['cell type']
Gene Metadata:  ['gene_names']
------------------------------------------------------------------------------
</code></pre></div></div>

<h2 id="create-a-list-of-batches">Create a list of batches</h2>

<p>To do batch integration, we have to create a Python list of batches.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Create a list of all the batches
</span>
<span class="n">batches</span> <span class="o">=</span> <span class="p">[</span><span class="n">bat1</span><span class="p">,</span> <span class="n">bat2</span><span class="p">]</span>
</code></pre></div></div>

<h2 id="integrate-and-merge-the-batches">Integrate and merge the batches</h2>

<p>The next step is to find common genes across the batches before merging them all into one dataset. The datasets can have different number and type of genes and cells. The <code class="language-plaintext highlighter-rouge">IntegrateBatches</code> function will attempt to find common genes across the datasets and merge the datasets into one SingleCell object. If no common genes are found a message will be displayed and the function will exit. We need to pass the batches list into this function as the first argument. The second argument is the name of the data column which stores the gene names in the SingleCell object. This needs to be passed as alist of string with the length same as the number of datasets as each dataset can store the gene name information under different names. We can also call <code class="language-plaintext highlighter-rouge">print()</code> on the integrated dataset to see a summary. We can see that another data column, ‘batch’, has been added to cell metadata to include batch information.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Integrate batches and merge into 1 SingleCell dataset
</span>
<span class="n">batches</span> <span class="o">=</span> <span class="n">IntegrateBatches</span><span class="p">(</span><span class="n">batches</span><span class="p">,</span> <span class="n">name_by</span> <span class="o">=</span> <span class="p">[</span><span class="s">'gene_names'</span><span class="p">,</span> <span class="s">'gene_names'</span><span class="p">])</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Integrating Batches . . .
Number of common genes in all batches:  100
Merging Batches . . .
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">batches</span><span class="p">.</span><span class="k">print</span><span class="p">()</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>------------------------------------------------------------------------------
Dataset:  simulation1+simulation2
------------------------------------------------------------------------------
Dimension:  (100, 2000)
Cell Metadata:  ['cell type' 'batch']
Gene Metadata:  ['gene_names']
------------------------------------------------------------------------------
</code></pre></div></div>

<h2 id="visualise-batches-before-correctionalignment">Visualise batches before correction/alignment</h2>

<p>We can visualize the integrated datasets before performation correction. We use the <code class="language-plaintext highlighter-rouge">tSNEPlot</code> function to generate a 2D scatter plot. We can see that the same cell types in different batches do not appear together or are not aligned.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">fig1</span> <span class="o">=</span> <span class="n">tSNEPlot</span><span class="p">(</span><span class="n">batches</span><span class="p">,</span>
                <span class="n">marker_by</span> <span class="o">=</span> <span class="s">'batch'</span><span class="p">,</span>
                <span class="n">color_by</span> <span class="o">=</span> <span class="s">'cell type'</span><span class="p">,</span>
                <span class="n">tsne_init</span> <span class="o">=</span> <span class="s">'pca'</span><span class="p">)</span>
<span class="n">fig1</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<p><img src="/assets/images/batfig1.png" alt="Fig 1" /></p>

<h2 id="run-batch-correction">Run batch correction</h2>

<p>To run batch correction, we use the <code class="language-plaintext highlighter-rouge">CorrectBatches</code> function. The first argument is the integrated batches as one SingleCell object. The second argument is <code class="language-plaintext highlighter-rouge">correct_order</code>, were the user can specify the order of the batches to use for correction. The <code class="language-plaintext highlighter-rouge">CorrectBatches</code> function will assign the first batch in the list as the reference batch and the second as the target batch. It will correct two batches at a time, and merge the first two to be used as the reference batch for correcting the third batch and so on. In this case, we only have two batches, so ‘simulation1’ will be the reference batch and ‘simulation2’ will be the target batch. The next parameter is the <code class="language-plaintext highlighter-rouge">sigma</code> parameter for the Gaussian kernel which is used to smooth correction vectors. The last parameter is <code class="language-plaintext highlighter-rouge">svd_dim</code>, the number of eigenvectors to select for checking the orthogonality assumption in MNN batch correction. For more info, run <code class="language-plaintext highlighter-rouge">help(CorrectBatches)</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">batches</span> <span class="o">=</span> <span class="n">CorrectBatches</span><span class="p">(</span><span class="n">batches</span><span class="p">,</span>
                         <span class="n">correct_order</span> <span class="o">=</span> <span class="p">[</span><span class="s">'simulation1'</span><span class="p">,</span> <span class="s">'simulation2'</span><span class="p">],</span> 
                         <span class="n">sigma</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
                         <span class="n">svd_dim</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Correcting batches &lt; simulation1  and  simulation2 &gt;
Merging Batches . . .
</code></pre></div></div>

<h2 id="visualize-correctedaligned-batches">Visualize corrected/aligned batches</h2>

<p>Finally, we can do <code class="language-plaintext highlighter-rouge">tSNEPlot</code> again to visualize the corrected batches.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">fig2</span> <span class="o">=</span> <span class="n">tSNEPlot</span><span class="p">(</span><span class="n">batches</span><span class="p">,</span>
                <span class="n">color_by</span> <span class="o">=</span> <span class="s">'cell type'</span><span class="p">,</span>
                <span class="n">marker_by</span> <span class="o">=</span> <span class="s">'batch'</span><span class="p">,</span>
                <span class="n">tsne_init</span> <span class="o">=</span> <span class="s">'pca'</span><span class="p">)</span>
<span class="n">fig2</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div>

<p><img src="/assets/images/batfig2.png" alt="Fig 2" /></p>

        
      </section>

      <footer class="page__meta">
        
        


        
  <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2020-09-05T00:00:00+12:00">September 5, 2020</time></p>


      </footer>

      <section class="page__share">
  

  <a href="https://twitter.com/intent/tweet?text=Performing+batch+correction+using+FEATS%20http%3A%2F%2Flocalhost%3A4000%2Fperforming-batch-correction-using-feats%2F" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=http%3A%2F%2Flocalhost%3A4000%2Fperforming-batch-correction-using-feats%2F" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A4000%2Fperforming-batch-correction-using-feats%2F" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="/welcome-to-my-site/" class="pagination--pager" title="Welcome to my site!
">Previous</a>
    
    
      <a href="/performing-clustering-using-feats/" class="pagination--pager" title="Performing clustering using FEATS
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
    <div class="page__related">
      <h4 class="page__related-title">You May Also Enjoy</h4>
      <div class="grid__wrapper">
        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/performing-outlier-detection-using-feats/" rel="permalink">Performing outlier detection using FEATS
</a>
      
    </h2>
    


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        3 minute read
      
    

  </p>

    <p class="archive__item-excerpt" itemprop="description">In this notebook, we perform outlier analysis using FEATS on mouse intestine dataset.

</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/performing-clustering-using-feats/" rel="permalink">Performing clustering using FEATS
</a>
      
    </h2>
    


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        5 minute read
      
    

  </p>

    <p class="archive__item-excerpt" itemprop="description">This notebook shows how to perform clustering using the FEATS package.

</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/welcome-to-my-site/" rel="permalink">Welcome to my site!
</a>
      
    </h2>
    


  <p class="page__meta">

    

    

    
      
      

      <i class="far fa-clock" aria-hidden="true"></i>
      
        less than 1 minute read
      
    

  </p>

    <p class="archive__item-excerpt" itemprop="description">Hello and welcome to my site! My name is Edwin Vans. I’m an academic who loves to write code and work on software and hardware projects. Stay tuned to my blo...</p>
  </article>
</div>

        
      </div>
    </div>
  
  
</div>

    </div>

    
      <div class="search-content">
        <div class="search-content__inner-wrap"><form class="search-content__form" onkeydown="return event.key != 'Enter';">
    <label class="sr-only" for="search">
      Enter your search term...
    </label>
    <input type="search" id="search" class="search-input" tabindex="-1" placeholder="Enter your search term..." />
  </form>
  <div id="results" class="results"></div></div>

      </div>
    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    

    
      
        
      
        
      
        
          <li><a href="https://github.com/edwinv87" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
        
      
        
      
        
      
        
      
    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2020 Edwin Vans. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>




<script src="/assets/js/lunr/lunr.min.js"></script>
<script src="/assets/js/lunr/lunr-store.js"></script>
<script src="/assets/js/lunr/lunr-en.js"></script>







  </body>
</html>
